File: Codeforces/Codeforces Round #787/D_Vertical_Paths.cpp
#include "bits/stdc++.h"
using namespace std;

void solve(){
    int n;
    cin >> n;

    int root;
    vector<vector<int>> children(n+1);

    for(int child = 1; child <= n; ++child){
        int parent;
        cin >> parent;
        if(parent == child) {
            root = parent;
        } else {
            children[parent].push_back(child);
        }
    }

    vector<vector<int>> answer;
    vector<int> path;

    function<void(int)> dfs = [&] (int node) {
        path.push_back(node);
        if(children[node].empty()){
            answer.push_back(path);
            path.clear();
        }
        for(int child : children[node]){
            dfs(child);
        }
    };

    dfs(root);

    cout << answer.size() << '\n';
    for(vector<int> path : answer){
        cout << path.size() << '\n';
        for(int node : path){
            cout << node << " ";
        }
        cout << '\n';
    }
    cout << '\n';
}

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);

    int t;
    cin >> t;
    while(t--){
        solve();
    }

}

File: Codeforces/Codeforces Round #787/D_Vertical_Paths.py
import sys

from types import GeneratorType
def bootstrap(f, stack=[]):
    def wrappedfunc(*args, **kwargs):
        if stack:
            return f(*args, **kwargs)
        else:
            to = f(*args, **kwargs)
            while True:
                if type(to) is GeneratorType:
                    stack.append(to)
                    to = next(to)
                else:
                    stack.pop()
                    if not stack:
                        break
                    to = stack[-1].send(to)
            return to
    return wrappedfunc

for _ in range(int(input())):
    n = int(input())
    graph = [[] for _ in range(n+1)]
    parents = list(map(int, input().split()))

    root = 1

    for i, p in enumerate(parents, 1):
        if i == p:
            root = i
        else:
            graph[p].append(i)

    path = []
    answer = []

    @bootstrap
    def dfs(node):
        path.append(node)
        if not graph[node]:
            answer.append(list(path))
            path.clear()
        for child in graph[node]:
            yield dfs(child)
        yield

    dfs(root)

    print(len(answer))
    for path in answer:
        print(len(path))
        print(*path)
    print()
